### PR TITLE
[#107] 사용자는 한 영화관에서 특정 날짜에 상영하는 영화 정보를 확인한다

### PR DESCRIPTION
**Pull request by yanJuicy**
## 개요
- 영화관, 날짜 요청 -> 영화 정보 조회 구현

## 추가기능
- 영화관 ID, 특정 날짜를 요청하면 영화관과 날짜에 맞는 영화 리스트를 반환

## 기타
- restdocds
<h3 id="_영화관과_상영_날짜를_요청하여_영화_리스트_가져오기" style="box-sizing: border-box; margin: 1em 0px 0.5em; padding: 0px; font-family: &quot;Open Sans&quot;, &quot;DejaVu Sans&quot;, sans-serif; font-weight: 300; font-style: normal; color: rgb(204, 204, 204); text-rendering: optimizelegibility; line-height: 1.2; font-size: 1.375em; word-spacing: -0.05em; font-variant-ligatures: normal; font-variant-caps: normal; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; -webkit-text-stroke-width: 0px; background-color: rgb(69, 73, 74); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">영화관과 상영 날짜를 요청하여 영화 리스트 가져오기</h3><div class="listingblock has-source-line data-line-/Users/hanju/IdeaProjects/BE-02-MarBox/build/generated-snippets/schedule-get-movie-by-date-and-theater/http-request.adoc-2" style="box-sizing: border-box; margin: 0px 0px 1.25em; padding: 0px; color: rgb(187, 187, 187); font-family: &quot;Noto Serif&quot;, &quot;DejaVu Serif&quot;, serif; font-size: 16px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(69, 73, 74); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><div class="title" style="box-sizing: border-box; margin: 0px 0px 0.25em; padding: 0px; line-height: 1.45; color: rgb(122, 37, 24); font-weight: 400; text-rendering: optimizelegibility; text-align: left; font-family: &quot;Noto Serif&quot;, &quot;DejaVu Serif&quot;, serif; font-size: 1rem; font-style: italic; filter: invert(90%) hue-rotate(180deg);">Request</div><div class="content" style="box-sizing: border-box; margin: 0px; padding: 0px; position: relative;"><pre class="CodeRay highlight nowrap" style="box-sizing: border-box; font-family: &quot;DejaVu Sans Mono&quot;, monospace; font-size: 0.8125em; white-space: pre; margin: 0px; padding: 1em; color: rgba(0, 0, 0, 0.9); line-height: 1.45; text-rendering: optimizespeed; overflow-wrap: normal; background: rgb(247, 247, 248); filter: invert(90%) hue-rotate(180deg); border-radius: 4px; overflow-x: auto;"><code data-lang="http" style="box-sizing: border-box; font-family: &quot;DejaVu Sans Mono&quot;, monospace; font-size: inherit; font-weight: 400; color: inherit; line-height: inherit; display: block;">GET /schedules?theaterId=1&amp;date=2022-06-30 HTTP/1.1
Host: localhost:8080</code></pre></div></div><div class="listingblock has-source-line data-line-/Users/hanju/IdeaProjects/BE-02-MarBox/build/generated-snippets/schedule-get-movie-by-date-and-theater/http-response.adoc-2" style="box-sizing: border-box; margin: 0px 0px 1.25em; padding: 0px; color: rgb(187, 187, 187); font-family: &quot;Noto Serif&quot;, &quot;DejaVu Serif&quot;, serif; font-size: 16px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(69, 73, 74); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><div class="title" style="box-sizing: border-box; margin: 0px 0px 0.25em; padding: 0px; line-height: 1.45; color: rgb(122, 37, 24); font-weight: 400; text-rendering: optimizelegibility; text-align: left; font-family: &quot;Noto Serif&quot;, &quot;DejaVu Serif&quot;, serif; font-size: 1rem; font-style: italic; filter: invert(90%) hue-rotate(180deg);">Response</div><div class="content" style="box-sizing: border-box; margin: 0px; padding: 0px; position: relative;"><pre class="CodeRay highlight nowrap" style="box-sizing: border-box; font-family: &quot;DejaVu Sans Mono&quot;, monospace; font-size: 0.8125em; white-space: pre; margin: 0px; padding: 1em; color: rgba(0, 0, 0, 0.9); line-height: 1.45; text-rendering: optimizespeed; overflow-wrap: normal; background: rgb(247, 247, 248); filter: invert(90%) hue-rotate(180deg); border-radius: 4px; overflow-x: auto;"><code data-lang="http" style="box-sizing: border-box; font-family: &quot;DejaVu Sans Mono&quot;, monospace; font-size: inherit; font-weight: 400; color: inherit; line-height: inherit; display: block;">HTTP/1.1 200 OK
Content-Type: application/json
X-Content-Type-Options: nosniff
X-XSS-Protection: 1; mode=block
Cache-Control: no-cache, no-store, max-age=0, must-revalidate
Pragma: no-cache
Expires: 0
X-Frame-Options: DENY
Content-Length: 391

{"movieList":[{"name":"영화1","limitAge":"CHILD","genre":"ACTION","runningTime":100,"posterImgLocation":"test/location"},{"name":"영화2","limitAge":"ADULT","genre":"ROMANCE","runningTime":120,"posterImgLocation":"test/location"},{"name":"영화3","limitAge":"CHILD","genre":"ANIMATION","runningTime":150,"posterImgLocation":"test/location"}],"theaterList":[],"dateList":[],"timeList":[]}</code></pre></div></div>

Path | Type | Description
-- | -- | --
movieList[] | Array | 영화 리스트
movieList[].name | String | 영화 이름
movieList[].limitAge | String | 영화 관람 등급
movieList[].genre | String | 장르
movieList[].runningTime | Number | 상영시간
movieList[].posterImgLocation | String | 포스터 이미지 경로
dateList[] | Array | 빈 배열
theaterList[] | Array | 빈 배열
timeList[] | Array | 빈 배열

### CODE REVIEW

File: src/main/java/prgrms/marco/be02marbox/domain/theater/controller/ScheduleController.java [hunk 1]
```diff
@@ -48,14 +49,18 @@ public ResponseEntity<List<ResponseFindMovie>> getCurrentMovieList() {
 	public ResponseEntity<ResponseFindSchedule> getSchedulesByRequestParam(
 		@RequestParam(required = false) Long movieId,
 		@RequestParam(required = false) Long theaterId,
-		@RequestParam(required = false) LocalDate date) {
+		@RequestParam(required = false) @DateTimeFormat(iso = DateTimeFormat.ISO.DATE) LocalDate date) {
```
Comment (Pawer0223) [inline src/main/java/prgrms/marco/be02marbox/domain/theater/controller/ScheduleController.java:52]: 오 포맷을 이렇게줄수도 있군요 👍

File: src/main/java/prgrms/marco/be02marbox/domain/theater/controller/ScheduleController.java [hunk 2]
```diff
@@ -48,14 +49,18 @@ public ResponseEntity<List<ResponseFindMovie>> getCurrentMovieList() {
...
+
+		ResponseFindSchedule responseFindSchedule = (new ResponseFindSchedule(Collections.emptyList(),
+			Collections.emptyList(), Collections.emptyList(), Collections.emptyList()));
```
Comment (Pawer0223) [inline src/main/java/prgrms/marco/be02marbox/domain/theater/controller/ScheduleController.java:55]: 해당코드는 항상 타는것 같은데 else로 분리해도 괜찮지 않을까요 ?

**Review by sdardew — APPROVED** [review]

File: src/main/java/prgrms/marco/be02marbox/domain/theater/service/ScheduleService.java [hunk 3]
```diff
@@ -97,6 +98,26 @@ public ResponseFindSchedule findMovieListAndDateListByTheaterId(Long theaterId)
 		return new ResponseFindSchedule(movieList, Collections.emptyList(), dateList, Collections.emptyList());
 	}
 
+	@Transactional(readOnly = true)
+	public ResponseFindSchedule findMovieListByTheaterIdAndDate(Long theaterId, LocalDate date) {
+		theaterRepository.findById(theaterId)
+			.orElseThrow(() -> new EntityNotFoundException(Message.INVALID_THEATER_EXP_MSG.getMessage()));
+		if (LocalDate.now().plusDays(CURRENT_SCHEDULE_PERIOD).isBefore(date) || LocalDate.now().isAfter(date)) {
```
Comment (Pawer0223) [inline src/main/java/prgrms/marco/be02marbox/domain/theater/service/ScheduleService.java:105]: 이 조건이 입력 변수 date를 검증하는것 같은데 findById보다 위에있으면 DB접근을 줄일 수 있지 않을까 합니다 !

File: src/main/java/prgrms/marco/be02marbox/domain/theater/repository/ScheduleRepository.java [hunk 4]
```diff
@@ -16,4 +16,10 @@ public interface ScheduleRepository extends JpaRepository<Schedule, Long> {
 		+ "WHERE CAST(s.startTime AS LocalDate) BETWEEN :startDay AND :endDay")
 	List<Schedule> findSchedulesBetweenStartDateAndEndDate(LocalDate startDay, LocalDate endDay);
 
+	@Query("SELECT s FROM Schedule s "
+		+ "JOIN FETCH s.theaterRoom "
+		+ "JOIN FETCH s.movie "
+		+ "WHERE CAST(s.startTime AS LocalDate) = :date")
+	List<Schedule> findScheduleByDate(LocalDate date);
```
Comment (Pawer0223) [inline src/main/java/prgrms/marco/be02marbox/domain/theater/repository/ScheduleRepository.java:23]: 👍

File: src/main/java/prgrms/marco/be02marbox/domain/theater/service/ScheduleService.java [hunk 5]
```diff
@@ -97,6 +98,26 @@ public ResponseFindSchedule findMovieListAndDateListByTheaterId(Long theaterId)
...
+			throw new DateTimeException(Message.INVALID_DATE_EXP_MSG.getMessage());
+		}
+
+		Set<TheaterRoom> theaterRooms = theaterRoomRepository.findAllByTheaterId(theaterId);
+		List<Schedule> schedulesOfDate = scheduleRepository.findScheduleByDate(date);
+
+		List<ResponseFindMovie> movieList = schedulesOfDate.stream()
+			.filter(schedule -> theaterRooms.contains(schedule.getTheaterRoom()))
```
Comment (Pawer0223) [inline src/main/java/prgrms/marco/be02marbox/domain/theater/service/ScheduleService.java:113]: Set의 contains로 쉽게 잘 풀어내셨군요.. 👍

File: src/main/java/prgrms/marco/be02marbox/domain/theater/controller/ScheduleController.java [hunk 6]
```diff
@@ -48,14 +49,18 @@ public ResponseEntity<List<ResponseFindMovie>> getCurrentMovieList() {
 	public ResponseEntity<ResponseFindSchedule> getSchedulesByRequestParam(
 		@RequestParam(required = false) Long movieId,
```
Comment (Pawer0223) [inline src/main/java/prgrms/marco/be02marbox/domain/theater/controller/ScheduleController.java:50]: movieId를 받을 필요가 있나요 ?? 아래에서 모두 movieId == null 조건만 걸려있는것 같아서요..
나중에 영화조건도 함께 받아서 처리하기위함인가요 ?

File: src/docs/asciidoc/schedule.adoc [hunk 7]
```diff
@@ -23,11 +23,20 @@ include::{snippets}/schedule-get-current-movies/http-request.adoc[]
 include::{snippets}/schedule-get-current-movies/http-response.adoc[]
 include::{snippets}/schedule-get-current-movies/response-fields.adoc[]
 
-### 한 영화관의 영화, 날짜 정보 가져오기
+### 영화관을 요청하여 영화 리스트, 상영 날짜 리스트 가져오기
```
Comment (Pawer0223) [inline src/docs/asciidoc/schedule.adoc:26]: 오히려 변경 전 이름이 더 이해하기 편할것 같다는 의견입니다 ㅎ

**Review by Pawer0223 — APPROVED** [review]
고생하셨습니다 !
크리티컬한 문제는 없어보여 approve하였습니다.

File: src/main/java/prgrms/marco/be02marbox/domain/theater/controller/ScheduleController.java (hunk 1 above)
Comment (yanJuicy) [reply src/main/java/prgrms/marco/be02marbox/domain/theater/controller/ScheduleController.java:52]: 날짜가 RequestParam에서는 String으로 전달이 와서 LocalDate 형식으로 변환이 필요했습니다.

File: src/main/java/prgrms/marco/be02marbox/domain/theater/controller/ScheduleController.java (hunk 2 above)
Comment (yanJuicy) [reply src/main/java/prgrms/marco/be02marbox/domain/theater/controller/ScheduleController.java:55]: 넵 그게 좋은거 같아요

File: src/main/java/prgrms/marco/be02marbox/domain/theater/controller/ScheduleController.java (hunk 6 above)
Comment (yanJuicy) [reply src/main/java/prgrms/marco/be02marbox/domain/theater/controller/ScheduleController.java:50]: 넵 movieId, theaterId, date 모두 파라미터로 받아 상영 시간을 확인할 때 필요합니다.

File: src/docs/asciidoc/schedule.adoc (hunk 7 above)
Comment (yanJuicy) [reply src/docs/asciidoc/schedule.adoc:26]: 앗 그런가요..
이름 짓는게 넘 어렵네요

Comment (Pawer0223) [reply src/docs/asciidoc/schedule.adoc:26]: 그냥 제 의견일뿐입니다 ㅎ
개발자 입장에서는 변경하신게 더 직관적이지만, API를 보는 사람입장에서는 이전이 낫다고 생각했습니다 !

**Pull request merged**